.github/workflows/ci-py.yml

```
@@ -68,8 +71,59 @@ jobs:
       - name: Lint with ruff
         run: poetry run ruff check
 
+  build_binary:
+    needs: changes
+    if: ${{ needs.changes.outputs.python == 'true' }}
+
+    name: Build HUGR binary
+    runs-on: ubuntu-latest
+    env:
+      SCCACHE_GHA_ENABLED: "true"
+      RUSTC_WRAPPER: "sccache"
+
+    steps:
+      - uses: actions/checkout@v4
+      - uses: mozilla-actions/sccache-action@v0.0.4
+      - name: Install stable toolchain
+        uses: dtolnay/rust-toolchain@stable
+      - name: Build HUGR binary
+        run: cargo build -p hugr-cli
+      - name: Upload the binary to the artifacts
+        uses: actions/upload-artifact@v4
+        with:
+          name: hugr_binary
+          path: target/debug/hugr
+  test:
+    needs: [changes, check, build_binary]
```

No need to wait for the `check` step to run this one.
```suggestion
    needs: [changes, build_binary]
```